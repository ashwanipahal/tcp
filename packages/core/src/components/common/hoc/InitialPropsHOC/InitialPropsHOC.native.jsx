import React from 'react';

/**
 * InitialPropsHOC
 * This is a HOC of Component that is wrapped in it
 * It calls getInitialProps method of wrapped component on load
 * and calls same getInitialProps method on brand switch
 *
 * @param {*} Component
 * @returns
 */
const InitialPropsHOC = Component => {
  return class InitialProps extends React.PureComponent<Props> {
    /**
     * @function componentDidMount
     * called when component is mount and calls getInitialProps method of wrapped component
     * and adds didFocus listener to the view which is called every time view is displayed
     *
     */
    componentDidMount() {
      this.refreshOnNextRender = true;
      this.addDidFocusListener();
    }

    UNSAFE_componentWillReceiveProps(nextProps) {
      const { appType } = nextProps;
      const { appType: prevAppType } = this.props;

      if (appType && appType !== prevAppType) {
        this.refreshOnNextRender = true;
      }
    }

    componentWillUnmount() {
      // Remove the did focus listener
      this.didFocusSubscription.remove();
    }

    /**
     * @function addDidFocusListener
     * Adds didFocus listener to view which is called every time view is displayed
     * It calls componentDidMount method of wrapped component if brand is switched
     * getInitialProps is called every time when component is focused and visible
     *
     */
    addDidFocusListener = () => {
      const { navigation } = this.props;
      if (!navigation.addListener) return;
      this.didFocusSubscription = navigation.addListener('didFocus', () => {
        if (this.refreshOnNextRender && this.component && this.component.componentDidMount) {
          this.component.componentDidMount();
        }

        if (this.component && this.component.getInitialProps) this.component.getInitialProps();
        this.refreshOnNextRender = false;
      });
    };

    /**
     * renders wrapped component
     *
     * @returns
     */
    render() {
      return (
        <Component
          {...this.props}
          ref={ref => {
            this.component = ref;
          }}
        />
      );
    }
  };
};

export default InitialPropsHOC;
