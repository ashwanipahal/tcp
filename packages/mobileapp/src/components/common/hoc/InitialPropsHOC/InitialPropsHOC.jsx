import React from 'react';

/**
 * InitialPropsHOC
 * This is a HOC of Component that is wrapped in it
 * It calls getInitialProps method of wrapped component on load
 * and calls same getInitialProps method on brand switch
 *
 * @param {*} Component
 * @returns
 */
const InitialPropsHOC = Component => {
  return class InitialProps extends React.PureComponent<Props> {
    /**
     * @function componentDidMount
     * called when component is mount and calls getInitialProps method of wrapped component
     * and adds didFocus listener to the view which is called every time view is displayed
     *
     */
    componentDidMount() {
      this.addDidFocusListener();
    }

    componentWillUnmount() {
      // Remove the did focus listener
      this.didFocusSubscription.remove();
    }

    /**
     * @function addDidFocusListener
     * Adds didFocus listener to view which is called every time view is displayed
     * It calls getInitialProps method of wrapped component if brand is switched
     *
     */
    addDidFocusListener = () => {
      const { navigation } = this.props;
      if (!navigation.addListener) return;
      this.didFocusSubscription = navigation.addListener('didFocus', () => {
        if (this.component && this.component.getInitialProps) this.component.getInitialProps();
      });
    };

    /**
     * renders wrapped component
     *
     * @returns
     */
    render() {
      return (
        <Component
          {...this.props}
          ref={ref => {
            this.component = ref;
          }}
        />
      );
    }
  };
};

export default InitialPropsHOC;
