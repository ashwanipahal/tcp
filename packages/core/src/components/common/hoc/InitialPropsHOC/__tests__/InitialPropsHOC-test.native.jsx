import React from 'react';
import { shallow } from 'enzyme';
import InitialPropsHOC from '../InitialPropsHOC.native';
import BagPageView from '../../../../features/CnC/BagPage/views/BagPage.view.native';

describe('InitialPropsHOC testcases', () => {
  let wrapper;
  const addListener = jest.fn();
  const BagPage = InitialPropsHOC(BagPageView);

  beforeEach(() => {
    wrapper = shallow(<BagPage navigation={{ addListener }} />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
