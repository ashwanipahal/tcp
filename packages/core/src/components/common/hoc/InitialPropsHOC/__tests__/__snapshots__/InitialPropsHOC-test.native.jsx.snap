// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InitialPropsHOC testcases should render correctly 1`] = `
<InitialProps
  navigation={
    Object {
      "addListener": [MockFunction] {
        "calls": Array [
          Array [
            "didFocus",
            [Function],
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
    }
  }
/>
`;
